File: projects/crawler/src/basango/services/crawler/async/tasks.py
from __future__ import annotations

import logging
from typing import Any

from basango.core.config import CrawlerConfig
from basango.core.config_manager import ConfigManager
from basango.domain import DateRange, PageRange, SourceKind, UpdateDirection
from basango.services import CsvPersistor
from basango.services.crawler.html_crawler import HtmlCrawler
from basango.services.crawler.wordpress_crawler import WordpressCrawler

from .queue import QueueManager, QueueSettings
from .schemas import (
    ArticleTaskPayload,
    ListingTaskPayload,
    ProcessedTaskPayload,
)


logger = logging.getLogger(__name__)


def schedule_async_crawl(
    *,
    source_id: str,
    env: str = "development",
    page_range: str | None = None,
    date_range: str | None = None,
    category: str | None = None,
    settings: QueueSettings | None = None,
):
    payload = ListingTaskPayload(
        source_id=source_id,
        env=env,
        page_range=page_range,
        date_range=date_range,
        category=category,
    )
    manager = QueueManager(settings=settings)
    job = manager.enqueue_listing(payload)
    logger.info("Scheduled listing collection job %s for source %s", job.id, source_id)
    return job.id


def collect_listing(payload: dict[str, Any]) -> int:
    data = ListingTaskPayload.from_dict(payload)
    manager = ConfigManager()
    pipeline = manager.get(data.env)
    source = pipeline.sources.find(data.source_id)
    if source is None:
        logger.error("Unknown source id %s", data.source_id)
        return 0

    crawler_config = CrawlerConfig(
        source=source,
        page_range=PageRange.create(data.page_range) if data.page_range else None,
        date_range=DateRange.create(data.date_range) if data.date_range else None,
        category=data.category,
        notify=False,
        direction=UpdateDirection.FORWARD,
    )
    client_config = pipeline.fetch.client
    queue_manager = QueueManager()

    if source.source_kind == SourceKind.HTML:
        crawler = HtmlCrawler(crawler_config, client_config)
        queued = _collect_html_listing(crawler, data, queue_manager)
    elif source.source_kind == SourceKind.WORDPRESS:
        crawler = WordpressCrawler(crawler_config, client_config)
        queued = _collect_wordpress_listing(crawler, data, queue_manager)
    else:
        logger.warning(
            "Async crawling not supported for source kind %s", source.source_kind
        )
        queued = 0

    logger.info("Queued %s article detail jobs for source %s", queued, data.source_id)
    return queued


def collect_article(payload: dict[str, Any]) -> dict[str, Any] | None:
    data = ArticleTaskPayload.from_dict(payload)
    manager = ConfigManager()
    pipeline = manager.get(data.env)
    source = pipeline.sources.find(data.source_id)
    if source is None:
        logger.error("Unknown source id %s", data.source_id)
        return None

    crawler_config = CrawlerConfig(
        source=source,
        date_range=DateRange.create(data.date_range) if data.date_range else None,
        category=data.category,
        notify=False,
        direction=UpdateDirection.FORWARD,
    )

    source_identifier = getattr(source, "source_id", data.source_id) or data.source_id
    persistors = [
        CsvPersistor(
            data_dir=pipeline.paths.data,
            source_id=str(source_identifier),
        )
    ]

    queue_manager = QueueManager()

    if source.source_kind == SourceKind.HTML:
        article = _collect_html_article(
            HtmlCrawler(crawler_config, pipeline.fetch.client, persistors=persistors),
            data,
        )
    elif source.source_kind == SourceKind.WORDPRESS:
        article = _collect_wordpress_article(
            WordpressCrawler(
                crawler_config, pipeline.fetch.client, persistors=persistors
            ),
            data,
        )
    else:
        logger.warning(
            "Async crawling not supported for source kind %s", source.source_kind
        )
        article = None

    if article:
        queue_manager.enqueue_processed(
            ProcessedTaskPayload(
                source_id=data.source_id,
                env=data.env,
                article=article,
            )
        )
        logger.info(
            "Persisted article %s and forwarded to processed queue",
            article.get("link"),
        )

    return article


def forward_for_processing(payload: dict[str, Any]) -> dict[str, Any] | None:
    data = ProcessedTaskPayload.from_dict(payload)
    article = dict(data.article) if data.article is not None else None
    if article is None:
        logger.info(
            "Ready for downstream processing: source=%s (no article)", data.source_id
        )
        return None
    logger.info(
        "Ready for downstream processing: source=%s link=%s",
        data.source_id,
        article.get("link"),
    )
    return article


def _collect_html_listing(
    crawler: HtmlCrawler,
    payload: ListingTaskPayload,
    queue_manager: QueueManager,
) -> int:
    source = crawler.source
    selector = source.source_selectors.articles
    if not selector:
        logger.warning(
            "No article selector configured for HTML source %s",
            source.source_id,
        )
        return 0

    page_range = crawler.config.page_range or crawler.get_pagination()
    queued = 0

    for page in range(page_range.start, page_range.end + 1):
        page_url = crawler._build_page_url(page)
        try:
            soup = crawler.crawl(page_url, page)
        except Exception as exc:  # noqa: BLE001
            logger.exception("Failed to crawl page %s: %s", page_url, exc)
            continue

        for node in soup.select(selector):
            link = crawler._extract_link(node)
            if not link:
                continue
            queue_manager.enqueue_article(
                ArticleTaskPayload(
                    source_id=payload.source_id,
                    env=payload.env,
                    url=link,
                    date_range=payload.date_range,
                    category=payload.category,
                )
            )
            queued += 1

    return queued


def _collect_wordpress_listing(
    crawler: WordpressCrawler,
    payload: ListingTaskPayload,
    queue_manager: QueueManager,
) -> int:
    page_range = crawler.config.page_range or crawler.get_pagination()
    queued = 0

    for page in range(page_range.start, page_range.end + 1):
        endpoint = crawler._posts_endpoint(page)
        try:
            response = crawler.client.get(endpoint)
            articles = response.json()
        except Exception as exc:  # noqa: BLE001
            logger.exception("Failed to fetch WordPress page %s: %s", endpoint, exc)
            continue

        if not isinstance(articles, list):
            logger.warning("Unexpected WordPress payload type: %s", type(articles))
            continue

        for entry in articles:
            queue_manager.enqueue_article(
                ArticleTaskPayload(
                    source_id=payload.source_id,
                    env=payload.env,
                    url=entry.get("link"),
                    data=entry,
                    date_range=payload.date_range,
                    category=payload.category,
                )
            )
            queued += 1

    return queued


def _collect_html_article(
    crawler: HtmlCrawler,
    payload: ArticleTaskPayload,
) -> dict[str, Any] | None:
    if not payload.url:
        logger.warning("Missing article url for HTML source %s", payload.source_id)
        return None

    crawler._current_article_url = payload.url  # type: ignore[attr-defined]
    try:
        soup = crawler.crawl(payload.url)
    except Exception as exc:  # noqa: BLE001
        logger.exception("Failed to crawl article %s: %s", payload.url, exc)
        return None

    crawler.fetch_one(str(soup), crawler.config.date_range)
    crawler.completed(False)
    return None


def _collect_wordpress_article(
    crawler: WordpressCrawler,
    payload: ArticleTaskPayload,
) -> dict[str, Any] | None:
    if payload.data is None:
        logger.warning("Missing WordPress payload for source %s", payload.source_id)
        return None

    crawler.fetch_one(payload.data, crawler.config.date_range)
    crawler.completed(False)
    return None

File: basango/apps/crawler/src/services/crawler/async/tasks.ts
import { logger } from "@basango/logger";

import {
        ListingTaskPayloadSchema,
        ArticleTaskPayloadSchema,
        ProcessedTaskPayloadSchema,
        ListingTaskPayload,
        ArticleTaskPayload,
        ProcessedTaskPayload,
} from "./schemas";
import {
	createQueueManager,
	QueueManager,
	QueueSettings,
	QueueSettingsInput,
} from "./queue";

export interface CrawlerTaskHandlers {
	collectListing: (payload: ListingTaskPayload) => Promise<number> | number;
	collectArticle: (payload: ArticleTaskPayload) => Promise<unknown> | unknown;
	forwardForProcessing: (
		payload: ProcessedTaskPayload,
	) => Promise<unknown> | unknown;
}

const notImplemented = (name: keyof CrawlerTaskHandlers) => () => {
	throw new Error(`Crawler task handler '${name}' is not implemented`);
};

let handlers: CrawlerTaskHandlers = {
	collectListing: notImplemented("collectListing"),
	collectArticle: notImplemented("collectArticle"),
	forwardForProcessing: notImplemented("forwardForProcessing"),
};

export const registerCrawlerTaskHandlers = (
	overrides: Partial<CrawlerTaskHandlers>,
): void => {
	handlers = { ...handlers, ...overrides };
};

export interface ScheduleAsyncCrawlOptions {
	sourceId: string;
	env?: string;
	pageRange?: string | null;
	dateRange?: string | null;
	category?: string | null;
	settings?: QueueSettings | QueueSettingsInput;
	queueManager?: QueueManager;
}

export const scheduleAsyncCrawl = async ({
        sourceId,
        env = "development",
        pageRange,
        dateRange,
        category,
        settings,
        queueManager,
}: ScheduleAsyncCrawlOptions): Promise<string> => {
        const payload = ListingTaskPayloadSchema.parse({
                source_id: sourceId,
                env,
                page_range: pageRange ?? undefined,
                date_range: dateRange ?? undefined,
                category: category ?? undefined,
        });

        const manager = queueManager ?? createQueueManager({ settings });
        logger.debug(
                {
                        sourceId,
                        env: payload.env,
                        pageRange: payload.page_range,
                        dateRange: payload.date_range,
                        category: payload.category,
                },
                "Scheduling listing collection job",
        );
        try {
                const job = await manager.enqueueListing(payload);
                logger.info(
                        { jobId: job.id, sourceId, env: payload.env },
                        "Scheduled listing collection job",
                );
                return job.id;
        } finally {
                if (!queueManager) {
                        await manager.close();
                }
	}
};

export const collectListing = async (payload: unknown): Promise<number> => {
        const data = ListingTaskPayloadSchema.parse(payload);
        logger.debug(
                {
                        sourceId: data.source_id,
                        env: data.env,
                        pageRange: data.page_range,
                        dateRange: data.date_range,
                        category: data.category,
                },
                "Collecting listing",
        );

        const result = await handlers.collectListing(data);
        const count = typeof result === "number" ? result : 0;

        logger.info(
                {
                        sourceId: data.source_id,
                        env: data.env,
                        queuedArticles: count,
                },
                "Listing collection completed",
        );

        return count;
};

export const collectArticle = async (payload: unknown): Promise<unknown> => {
        const data = ArticleTaskPayloadSchema.parse(payload);
        logger.debug(
                {
                        sourceId: data.source_id,
                        env: data.env,
                        url: data.url,
                        page: data.page,
                },
                "Collecting article",
        );

        const result = await handlers.collectArticle(data);

        logger.info(
                {
                        sourceId: data.source_id,
                        env: data.env,
                        url: data.url,
                },
                "Article collection completed",
        );

        return result;
};

export const forwardForProcessing = async (
        payload: unknown,
): Promise<unknown> => {
        const data = ProcessedTaskPayloadSchema.parse(payload);
        logger.debug(
                {
                        sourceId: data.source_id,
                        env: data.env,
                },
                "Forwarding article for processing",
        );

        const result = await handlers.forwardForProcessing(data);

        logger.info(
                {
                        sourceId: data.source_id,
                        env: data.env,
                },
                "Article forwarded for processing",
        );

        return result;
};
